design/volume-group-snapshot.md

```diff
@@ -0,0 +1,504 @@
+    PVC3 --- PVC4
+    PVC3 -- "group: B" --> VGS_B[VGS-B group: B]
+    PVC4 -- "group: B" --> VGS_B
+```
+
+3. Pod Mounts: Pod1 mounts both PVC1 and PVC2, Pod2 mounts PVC1 and PVC3.  
+   Grouping:   
+     Group A: PVC1 and PVC2  
+     Group B: PVC3  
+   ItemBlock: All objects-Pod1, Pod2, PVC1, PVC2, and PVC3, are collected into a single item block.  
+   VolumeGroupSnapshots:  
+     PVC1 and PVC2 (group A) point to the same VGS (VGS (group: A)).   
+     PVC3 (group B) point to a different VGS (VGS (group: B)).  
+```mermaid
+flowchart TD
+    subgraph ItemBlock
+        P1[Pod1]
+        P2[Pod2]
+        PVC1[PVC1 group: A]
+        PVC2[PVC2 group: A]
+        PVC3[PVC3 group: B]
+    end
+
+    %% Pod mount relationships
+    P1 -->|mounts| PVC1
+    P1 -->|mounts| PVC2
+    P2 -->|mounts| PVC1
+    P2 -->|mounts| PVC3
+
+    %% Grouping for Group A: PVC1 and PVC2 are grouped into VGS_A
+    PVC1 --- PVC2
+    PVC1 -- "Group A" --> VGS_A[VGS Group A]
+    PVC2 -- "Group A" --> VGS_A
+
+    %% Grouping for Group B: PVC3 grouped into VGS_B
+    PVC3 -- "Group B" --> VGS_B[VGS Group B]
+    
+```
+
+#### Updates to CSI PVC plugin:
+  - This is the plugin which creates VolumeSnapshots for the CSI PVCs if required.
+  - We might encounter the following cases:
+    - PVC has VGS label and VGS does not exist:
+      - We can check if VGS exists by listing the VGS in the namespace using the backup UUID 
+        along with the label used for the VGS (we will be creating VGS with backup UUID and user specified VGS  group label) and see if the PVC is part of any of the existing VGS for the particular backup operation we are currently in.
+      - If not then create the VGS object with the required labels (in this case backup UUID and the user specified VGS group label) and also add the PVC names (that are included in the VGS) as 
+        annotations.
+      - Once VGS is created, wait for VGS object to be `readyToUse`, it implies that all the related VS/VSC objects are ready to use too.
+      - Add the VGS object as an additional item.
+      - For non-datamover case:
+        - Here, we need to skip the creation of VS as the VS object get created via creation of VGS object.
+        - We just need to add the VS objects that got created via VGS object as additional items and the rest of the workflow remains the same for non-datamover cases.
+      - For datamover case:
+        - We need to skip creation of VS but need to create DataUploads for the VS instance that got created due to VGS creation.
+        - Fetch the correct VS instances that got created via VGS and is related to the current PVC.
+        - Finally create DataUpload object for this VS/PVC pair.
+        - Add the DataUpload as an additional item.
+    - PVC has VGS label and VGS already exists:
+      - In this case we do not want to re-create VGS object.
+      - The required VS/DU for the PVC already exists as VGS instance exists for the PVC and backup UUID.
```

Comment:
Add: wait for appearance of snapshotHandle

We still need to wait for the VSC in the right status before moving forwad

Reply:
Thats already present in current DM flow of CSI plugin